File: infrastructure/web/src/main/java/org/acme/infra/web/graphql/CustomerGraphQLQuery.java
package org.acme.infra.web.graphql;

import org.acme.infra.web.model.JsonCustomer;
import org.acme.app.query.GetCustomers;
import org.eclipse.microprofile.graphql.GraphQLApi;
import org.eclipse.microprofile.graphql.Query;
import org.eclipse.microprofile.graphql.Description;
import java.util.stream.Collectors;
import java.util.List;


@GraphQLApi
public class CustomerGraphQLQuery {

    private GetCustomers getCustomers;

    public CustomerGraphQLQuery(GetCustomers getCustomers) {
        this.getCustomers = getCustomers;
    }

    @Query("customers")
    @Description("Get all customers")
    public List<JsonCustomer> getAllFilms() {
        return getCustomers.getCustomers().stream()
        .map(response -> new JsonCustomer(response.getName()))
        .collect(Collectors.toList());
    }
}

File: infrastructure/web/src/main/kotlin/org/acme/infra/web/graphql/CustomerGraphQLQuery.kt
package org.acme.infra.web.graphql

import org.acme.app.command.CreateCustomerCommand
import org.acme.app.query.GetCustomerQuery
import org.acme.app.query.GetCustomersQuery
import org.acme.infra.web.model.JsonCustomer
import org.eclipse.microprofile.graphql.Description
import org.eclipse.microprofile.graphql.GraphQLApi
import org.eclipse.microprofile.graphql.Mutation
import org.eclipse.microprofile.graphql.Query
import java.util.stream.Collectors

@GraphQLApi
class CustomerGraphQLQuery(
    private val getCustomers: GetCustomersQuery,
    private val getCustomer: GetCustomerQuery,
    private val createCustomerCommand: CreateCustomerCommand,
) {

    @Description("Get all customers")
    @Query("customers")
    fun allCustomers(): List<JsonCustomer> {
        return getCustomers.customers {
            it.map { c -> JsonCustomer(c.id, c.name) }
        }
    }

    @Description("Create a customer")
    @Mutation("createCustomer")
    fun createCustomer(name: String): JsonCustomer {
        val id = createCustomerCommand.create(CreateCustomerCommand.Request(name)).id
        return getCustomer.customer(GetCustomerQuery.Request(id)) {
            JsonCustomer(it.id, it.name)
        }
    }
}
